processor/transformprocessor/factory.go

```diff
@@ -22,13 +25,67 @@ import (
 
 var processorCapabilities = consumer.Capabilities{MutatesData: true}
 
-func NewFactory() processor.Factory {
+type transformProcessorFactory struct {
+	additionalLogFunctions    []ottl.Factory[ottllog.TransformContext]
+	additionalSpanFunctions   []ottl.Factory[ottlspan.TransformContext]
+	additionalMetricFunctions []ottl.Factory[ottlmetric.TransformContext]
+}
+
+// FactoryOption applies changes to transformProcessorFactory.
+type FactoryOption func(factory *transformProcessorFactory)
+
+// defaultAdditionalMetricFunctions returns an empty slice of ottl function factories.
+func defaultAdditionalMetricFunctions() []ottl.Factory[ottlmetric.TransformContext] {
+	return []ottl.Factory[ottlmetric.TransformContext]{}
+}
+
+// defaultAdditionalLogFunctions returns an empty slice of ottl function factories.
+func defaultAdditionalLogFunctions() []ottl.Factory[ottllog.TransformContext] {
+	return []ottl.Factory[ottllog.TransformContext]{}
+}
+
+// defaultAdditionalSpanFunctions returns an empty slice of ottl function factories.
+func defaultAdditionalSpanFunctions() []ottl.Factory[ottlspan.TransformContext] {
+	return []ottl.Factory[ottlspan.TransformContext]{}
+}
+
+// WithAdditionalMetricFunctions adds ottl metric functions to resulting processor
+func WithAdditionalMetricFunctions(metricFunctions []ottl.Factory[ottlmetric.TransformContext]) FactoryOption {
+	return func(factory *transformProcessorFactory) {
+		factory.additionalMetricFunctions = metricFunctions
+	}
+}
+
+// WithAdditionalLogFunctions adds ottl log functions to resulting processor
+func WithAdditionalLogFunctions(logFunctions []ottl.Factory[ottllog.TransformContext]) FactoryOption {
+	return func(factory *transformProcessorFactory) {
+		factory.additionalLogFunctions = logFunctions
+	}
+}
+
+// WithAdditionalSpanFunctions adds ottl span functions to resulting processor
+func WithAdditionalSpanFunctions(spanFunctions []ottl.Factory[ottlspan.TransformContext]) FactoryOption {
+	return func(factory *transformProcessorFactory) {
+		factory.additionalSpanFunctions = spanFunctions
+	}
+}
```

Comment:
These options should probably live in `pkg/ottl` so they can be shared by multiple processors, instead of having to be duplicated in each processor that uses OTTL.

Reply:
I removed `pkg/ottl/ottlprocessor` based on the last response from @edmocosta regarding the complexity of creating a shared API that would work for all OTTL based components.

> Yes, I agree we should aim for consistent APIs, but I still don't think it should be enforced/defined by the OTTL package. It already provides the parsers/utilities that components need to use the language, but how they do that, is up to them.

> There are many different use-cases for OTTL at the moment (processor, connectors...), and it would be tricky to come up with some common options that perfectly fit all components. For example, providing the WithAdditionalSpanEventFunctions option for a processor that only works with logs wouldn't make much sense.